Installation issue: paraview

### Steps to reproduce the issue

Spack fails to build paraview on a compute node with no gpu.
```console
$ spack install paraview @6 -qt -x -cuda +mpi +python +osmesa_fallback ^mesa -glx
```


### Error message

<details><summary>Error message</summary>

<pre>
[x] bwhug3k paraview@6.1.0 failed

1 error found in paraview@6.1.0 build log:
-- lines 172 to 184 --
  -- Could NOT find freetype (missing: freetype_DIR)
  -- Found Freetype: /home/criann/gbarbier/spack-test/spack/opt/spack/linux-zen4/freetype-2.14.2-wpvs6vkwu3sngrnchanm5rell2kfit5j/lib/libfreetype.so (found version "2.14.2")
  -- Found JsonCpp: /home/criann/gbarbier/spack-test/spack/opt/spack/linux-zen4/jsoncpp-1.9.6-3x7352t7edmbfa55xayry3yzv4zh6wqt/lib64/libjsoncpp.so (found suitable version "1.9.6", minimum required is "0.7.0")
  -- Found MPI_C: /home/criann/gbarbier/spack-test/spack/opt/spack/linux-zen4/openmpi-5.0.10-7icyhweofxbje4tunp55hgdftavphog6/lib/libmpi.so (found version "3.1")
  -- Found MPI: TRUE (found version "3.1") found components: C
  -- Could NOT find X11 (missing: X11_X11_INCLUDE_PATH X11_X11_LIB)
> CMake Error at VTK/CMake/vtkModule.cmake:5400 (message):
    Could not find the X11 external dependency.
  Call Stack (most recent call first):
    VTK/Utilities/X11/CMakeLists.txt:3 (vtk_module_find_package)


  -- Configuring incomplete, errors occurred!
</pre></details>


### Information on your system

* **Spack:** 1.2.0.dev0 (https://github.com/spack/spack/commit/7e864787bd15a516314d86002ce1cbabde7cbbe9)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/d060bd9e1803b10c35af3ecab5eee9778f75d039
* **Python:** 3.9.21
* **Platform:** linux-rhel9-zen4

### Additional information

It seems to me that in this spec, paraview depends on `osmesa` but not on `gl`. The package decides to define `-DVTK_USE_X=OFF` when the spec satisfies `^[virtuals=gl] osmesa` which isn't the case.

[spack-build-env.txt](https://github.com/user-attachments/files/27015312/spack-build-env.txt)
[spack-build-out.txt](https://github.com/user-attachments/files/27015322/spack-build-out.txt)

maintainers: @danlipsa @kwryankrattiger @vicentebolea

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

--- a/repos/spack_repo/builtin/packages/paraview/package.py
+++ b/repos/spack_repo/builtin/packages/paraview/package.py
@@ -596,28 +596,23 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
                 return on
             return off
 
-        def use_x11():
-            """Return false if osmesa or egl are requested"""
-            if (
-                spec.satisfies("^[virtuals=gl] osmesa")
-                or spec.satisfies("^[virtuals=gl] egl")
-                or spec.satisfies("platform=windows")
-            ):
-                return "OFF"
-            return "ON"
-
-        rendering = variant_bool("+opengl2", "OpenGL2", "OpenGL")
         includes = variant_bool("+development_files")
 
         cmake_args = [
-            "-DVTK_OPENGL_HAS_OSMESA:BOOL=%s" % variant_bool("^[virtuals=gl] osmesa"),
-            "-DVTK_USE_X:BOOL=%s" % use_x11(),
             "-DPARAVIEW_INSTALL_DEVELOPMENT_FILES:BOOL=%s" % includes,
             "-DBUILD_TESTING:BOOL=OFF",
             "-DOpenGL_GL_PREFERENCE:STRING=LEGACY",
+            self.define_from_variant("VTK_USE_X", "x"),
             self.define_from_variant("PARAVIEW_ENABLE_VISITBRIDGE", "visitbridge"),
             self.define_from_variant("VISIT_BUILD_READER_Silo", "visitbridge"),
         ]
+
+        if spec.satisfies("@:5"):
+            cmake_args.append(
+                "-DVTK_OPENGL_HAS_OSMESA:BOOL=%s" % variant_bool("^[virtuals=gl] osmesa")
+            )
+        else:
+            cmake_args.append(self.define_from_variant("VTK_OPENGL_HAS_OSMESA", "osmesa_fallback"))
 
         if spec.satisfies("^[virtuals=gl] egl"):
             cmake_args.append("-DVTK_OPENGL_HAS_EGL:BOOL=ON")
@@ -673,6 +668,7 @@ class Paraview(CMakePackage, CudaPackage, ROCmPackage):
                 ]
             )
         else:
+            rendering = variant_bool("+opengl2", "OpenGL2", "OpenGL")
             cmake_args.extend(
                 [
                     "-DPARAVIEW_BUILD_EXAMPLES:BOOL=%s" % variant_bool("+examples"),

